### PR TITLE
Allow building on Node 7+

### PR DESCRIPTION
**Pull request by winstliu**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Removes a restriction that prevented people from trying to build Atom using Node 7+.  Now that apm has been updated this should work again.
In addition, I fixed a potential bug where `stdout` was being declared as a constant but then being reassigned.  I'm not sure this would work; even if it did, it's best that it's rewritten to be a `let` instead.

### Alternate Designs

None.

### Why Should This Be In Core?

Already in core.

### Benefits

More supported Node versions when building.

### Possible Drawbacks

None.

### Applicable Issues

None.

If CI is green, I will revert the CI changes and :boat:.

### CODE REVIEW

**Comment by 50Wliu** [comment]
Turns out that the linting script doesn't lint the `script` folder!  There were a bunch of warnings in there, some of which become full-blown errors when running under Node 7 and improved ES6 support.  So I've added that to the list of files to lint, and will lint all those files as a part of this PR.

I also noticed that specs written in JavaScript aren't being linted, but in the interest of keeping this PR small, I'll be opening a different PR for that.

**Comment by 50Wliu** [comment]
Ok, builds are green.  Though this PR turned out larger-than-expected, and I want to point out some stuff that I'm not sure I did right for review.

File: script/lib/code-sign-on-mac.js [hunk 1]
```diff
@@ -10,8 +10,9 @@ module.exports = function (packagedAppPath) {
     return
   }
 
+  // This declaration is outside of the try block because it is also used for the finally block
+  const certPath = path.join(os.tmpdir(), 'mac.p12')
```
Comment (50Wliu) [inline script/lib/code-sign-on-mac.js]: I'm reasonably sure this shouldn't ever throw as long as `os.tmpdir()` works.

File: script/lib/generate-metadata.js [hunk 2]
```diff
@@ -27,8 +27,8 @@ function buildBundledPackagesMetadata () {
     const packagePath = path.join(CONFIG.intermediateAppPath, 'node_modules', packageName)
     const packageMetadataPath = path.join(packagePath, 'package.json')
     const packageMetadata = JSON.parse(fs.readFileSync(packageMetadataPath, 'utf8'))
-    normalizePackageData(packageMetadata, () => {
-      throw new Error(`Invalid package metadata. ${metadata.name}: ${msg}`)
+    normalizePackageData(packageMetadata, (msg) => {
+      throw new Error(`Invalid package metadata. ${packageName}: ${msg}`)
```
Comment (50Wliu) [inline script/lib/generate-metadata.js]: `metadata.name` doesn't exist, but `packageName` does.  I'm not sure the two are equivalent without knowing what `metadata.name` was supposed to refer to.

File: script/lib/generate-metadata.js [hunk 3]
```diff
@@ -118,7 +118,7 @@ function checkDeprecatedPackagesMetadata () {
   for (let packageName of Object.keys(deprecatedPackagesMetadata)) {
     const packageMetadata = deprecatedPackagesMetadata[packageName]
     if (packageMetadata.version && !semver.validRange(packageMetadata.version)) {
-      throw new Error(`Invalid range: ${version} (${name}).`)
+      throw new Error(`Invalid range: ${packageMetadata.version} (${packageName}).`)
```
Comment (50Wliu) [inline script/lib/generate-metadata.js]: Ditto here.

File: script/lib/transpile-babel-paths.js [hunk 4]
```diff
@@ -6,15 +6,6 @@ const glob = require('glob')
...
 const CONFIG = require('../config')
-const BABEL_OPTIONS = require('../../static/babelrc.json')
-const BABEL_PREFIXES = [
-  "'use babel'",
-  '"use babel"',
-  '/** @babel */',
-  '/* @flow */'
-]
-const PREFIX_LENGTH = Math.max.apply(null, BABEL_PREFIXES.map(prefix => prefix.length))
-const BUFFER = Buffer(PREFIX_LENGTH)
```
Comment (50Wliu) [inline script/lib/transpile-babel-paths.js]: All these are unused, but seem kinda important.  I hope I'm not missing anything here?

File: script/test [hunk 5]
```diff
@@ -71,7 +71,7 @@ for (let packageName in CONFIG.appMetadata.packageDependencies) {
     console.log(`Executing ${packageName} tests`.bold.green)
     const cp = childProcess.spawn(executablePath, testArguments)
     let stderrOutput = ''
-    cp.stderr.on('data', data => stderrOutput += data)
+    cp.stderr.on('data', data => (stderrOutput += data))
```
Comment (50Wliu) [inline script/test]: Little "hack" to get around standard complaint.

File: spec/menu-manager-spec.coffee [hunk 6]
```diff
@@ -62,7 +62,7 @@ describe "MenuManager", ->
       atom.keymaps.add 'test', 'atom-workspace': 'ctrl-b': 'b'
       menu.update()
 
-      waits 1
+      waits 50
```
Comment (50Wliu) [inline spec/menu-manager-spec.coffee]: Appveyor failed twice on this spec so I bumped the waits to be consistent with the other tests in this file and now it passes.

**Comment by 50Wliu** [comment]
Note: While running `script/build` using Node 7 now completes successfully, that _does not_ mean that Atom behaves as expected when built using Node 7, nor is Node 7+ supported.

**Comment by torn4dom4n** [comment]
Why don't you use Node 6 for LTS?

**Comment by 50Wliu** [comment]
I'm not planning to use Node 7 :).  This PR just _allows_ for Atom to be built in Nice 7 instead of erroring.  I will be reverting the CI changes before merging.

File: script/lib/transpile-babel-paths.js (hunk 4 above)
Comment (BinaryMuse) [inline script/lib/transpile-babel-paths.js]: Looks like they're here now: https://github.com/atom/atom/blob/e4f2c80eb5f124012524297622f20c540bbe3822/src/babel.js

**Comment by BinaryMuse** [comment]
@50Wliu I plan on taking a closer look at this soon. Are you okay with me resolving the merge conflicts?

**Comment by 50Wliu** [comment]
Yeah, of course.

File: script/test (hunk 5 above)
Comment (thomasjo) [inline script/test]: Am I missing something here? Why not use curly brackets here? Surely the complaint is due to the assignment and return of the original statement?

Comment (50Wliu) [inline script/test]: Yes, you're right.  I've been using curly brackets in other places to fix the warning but not sure why I didn't here.

**Comment by adamreisnz** [comment]
Any progress on this? We will be getting Node 8 soon and at present Atom still can't build on Node 7.9.0.

**Comment by BinaryMuse** [comment]
Atom now builds successfully on Node 7 and Node 8! Thanks @50Wliu for taking care of this.

The Travis test is currently not running due to issues on Travis' infrastructure, but I feel confident in the changes here and it passes with Node 7, so let's get this into master. 🎉

**Comment by adamreisnz** [comment]
Great work, thanks! Just in time for Node 8 ;)

**Pull request merged**